src/main/java/org/folio/service/mapping/MappingServiceImpl.java

```diff
@@ -39,27 +46,47 @@
   public MappingServiceImpl() {
     this.ruleProcessor = new RuleProcessor();
     this.ruleFactory = new RuleFactory();
+    this.mappingThreadPool = Executors.newWorkStealingPool(MAPPING_POOL_SIZE);
   }
 
   @Override
   public List<String> map(List<JsonObject> instances, MappingProfile mappingProfile, String jobExecutionId, OkapiConnectionParams connectionParams) {
     if (CollectionUtils.isEmpty(instances)) {
       return Collections.emptyList();
     }
-    List<String> records = new ArrayList<>();
     ReferenceData referenceData = referenceDataProvider.get(jobExecutionId, connectionParams);
     List<Rule> rules = getRules(mappingProfile, connectionParams);
-    for (JsonObject instance : instances) {
-      String record = runMappingProcess(instance, referenceData, rules);
-      records.add(record);
+    return mapInstances(instances, referenceData, rules);
+  }
+
+  private List<String> mapInstances(List<JsonObject> instances, ReferenceData referenceData, List<Rule> rules) {
+    List<String> records = null;
+    try {
+      records = mappingThreadPool.submit(() -> instances.parallelStream()
+        .map(instance -> mapInstance(instance, referenceData, rules))
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .collect(Collectors.toList()))
+        .get();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```

Comment:
Do you think it might be worthwhile to add the instances in a debug log that failed to be mapped?

Reply:
Debug log to track failed instances to be mapped, is already in place, line 87